identity/src/main/java/com/android/identity/MobileSecurityObjectGenerator.java

```diff
@@ -0,0 +1,272 @@
+     * included in the <code>authorizedDataElements</code> mapping. If a namespace is included here,
+     * then it should not be included in {@link #setDeviceKeyAuthorizedNameSpaces(List)}
+     *
+     * @param authorizedDataElements A mapping from namespaces to a list of
+     *                               <code>DataElementIdentifier</code>
+     * @return The <code>MobileSecurityObjectGenerator</code>.
+     */
+    @NonNull
+    public MobileSecurityObjectGenerator setDeviceKeyAuthorizedDataElements(
+            @NonNull Map<String, List<String>> authorizedDataElements) {
+
+        mAuthorizedDataElements.clear();
+        mAuthorizedDataElements.putAll(authorizedDataElements);
+        return this;
+    }
+
+    /**
+     * Provides extra info for the mdoc authentication public key as part of the
+     * <code>KeyInfo</code> portion of the <code>DeviceKeyInfo</code>.
+     *
+     * @param keyInfo A mapping to represent additional key information.
+     * @return The <code>MobileSecurityObjectGenerator</code>.
+     */
+    @NonNull
+    public MobileSecurityObjectGenerator setDeviceKeyInfo(@NonNull Map<Integer, byte[]> keyInfo) {
+        mKeyInfo.clear();
+        mKeyInfo.putAll(keyInfo);
+        return this;
+    }
+
+    /**
+     * Sets the <code>ValidityInfo</code> structure which contains information related to the
+     * validity of the MSO and its signature. This must be called before generating since this a
+     * required component of the <code>MobileSecurityObject</code>.
+     *
+     * @param signed         The timestamp at which the MSO signature was created.
+     * @param validFrom      The timestamp before which the MSO is not yet valid. This shall be
+     *                       equal or later than the signed element.
+     * @param validUntil     The timestamp after which the MSO is no longer valid. This shall be
+     *                       later than the validFrom element.
+     * @param expectedUpdate Optional: if provided, represents the timestamp at which the issuing
+     *                       authority infrastructure expects to re-sign the MSO, else, null
+     * @return The <code>MobileSecurityObjectGenerator</code>.
+     * @exception IllegalArgumentException if the times are do not meet the constraints.
+     */
+    @NonNull
+    public MobileSecurityObjectGenerator setValidityInfo(@NonNull Timestamp signed,
+            @NonNull Timestamp validFrom, @NonNull Timestamp validUntil,
+            @Nullable Timestamp expectedUpdate) {
+
+        if (validFrom.toEpochMilli() < signed.toEpochMilli()) {
+            throw new IllegalArgumentException("The validFrom timestamp should be equal or later " +
+                    "than the signed timestamp");
+        }
+        if (validUntil.toEpochMilli() <= validFrom.toEpochMilli()) {
+            throw new IllegalArgumentException("The validUntil timestamp should be later " +
+                    "than the validFrom timestamp");
+        }
```

Comment:
I like that we are validation user input. I would including comments such as

```
  // 18013-5 Section 9.1.2.4 says "The timestamp of validFrom shall be equal or later than the signed element."
  if (validFrom.toEpochMilli() < signed.toEpochMilli()) {
    [...]
```

to point the reader to where the standard requires this.

Reply:
I also added user validation for the relationship btw `authorizedNameSpaces` and `authorizedDataElements`

Reply:
SG, thanks!